parse_node/parse_node_with_comments.ts

```diff
@@ -0,0 +1,180 @@
+import ts from "typescript"
+import {
+  ParseNodeType,
+  parseNodeWithoutComments,
+  ParseState,
+} from "../parse_node"
+import { Test } from "../tests/test"
+
+const getBackwardsClosestNewLinesCount = (text: string, pos: number) => {
+  let count = 0
+
+  for (let i = pos - 1; i >= 0; i--) {
+    const char = text.charAt(i)
+    if (char === "\t" || char === " ") {
+      continue
+    }
+
+    if (char === "\n") {
+      count++
+      continue
+    }
+
+    break
+  }
+
+  return count
+}
+
+export const parseNodeWithComments = (
```

Comment:
naming nit: It's a little hard for me to keep track of what parseNodeWithoutComments and parseNodeWithoutComments do. 

How about we rename `parseNodeWithoutComments` back to `parseNode`, and name `parseNodeWithComments` to `parseComments`? 

----

And less of a nit, more of a readability thing:

The earlier state logic of how methods got called was basically parseNode -> parseSomething -> parseNode -> parseSomethingElse etc - that was pretty easy to follow. I'm a bit worried that if we go down this route we'll add more conditionals into parseNode as we find more cases we need to handle top-level, and then it'll be harder to follow the state machine logic. 

Would it be possible to have parseNode go back to being the top level parse function, and have it conditionally call parseNodeWithComments? I realize it's a little tricky to do this as written, but it seemed important to me for code clarity.

Reply:
> Would it be possible to have parseNode go back to being the top level parse function, and have it conditionally call parseNodeWithComments? I realize it's a little tricky to do this as written, but it seemed important to me for code clarity.

It would be possible but instead of current nice looking switch with returns I would need to introduce a variable for saving the result and then process the comments for it. All single line returns in switch cases would be rewritten to variable assignment and break keyword. That is why old `parseNode` has new name `parseNodeWithoutComments`and `parseNode` was replaced with a wrapper. 

```ts
// old code
  switch (genericNode.kind) {
    case SyntaxKind.SourceFile:
      return parseSourceFile(genericNode as ts.SourceFile, props)
    case SyntaxKind.ImportDeclaration:
      return parseImportDeclaration(genericNode as ts.ImportDeclaration, props)
    case SyntaxKind.TypeReference:
      return parseTypeReference(genericNode as ts.TypeReferenceNode, props)
    case SyntaxKind.TypeAliasDeclaration:
      return parseTypeAliasDeclaration(
        genericNode as ts.TypeAliasDeclaration,
        props
      )
      // ...
```
```ts
// new code
  let result;

  switch (genericNode.kind) {
    case SyntaxKind.SourceFile:
      result = parseSourceFile(genericNode as ts.SourceFile, props); break
    case SyntaxKind.ImportDeclaration:
      result = parseImportDeclaration(genericNode as ts.ImportDeclaration, props); break
    case SyntaxKind.TypeReference:
      result = parseTypeReference(genericNode as ts.TypeReferenceNode, props); break
    case SyntaxKind.TypeAliasDeclaration:
      result = parseTypeAliasDeclaration(
        genericNode as ts.TypeAliasDeclaration,
        props
      ); break
      // ...

     // additional code for comment parsing...
`

Reply:
I changed it to:

```ts
export const parseNode = (
  genericNode: ts.Node,
  props: ParseState,
  options?: { ignoreComments?: boolean }
): ParseNodeType => {
  // TsGdProjectClass.Paths is undefined when running tests
  if (!(TsGdProjectClass.Paths ?? {}).removeComments && !options?.ignoreComments) {
    return parseComments(genericNode, props)
  }
  // ...
```

Is this okay?

Reply:
Yeah, that looks good to me!